pkg/policy/selectorcache.go

```diff
@@ -901,6 +929,15 @@ func (sc *SelectorCache) AddIdentitySelector(user CachedSelectionUser, selector
 		newIDSel.namespaces = namespaces
 	}
 
+	// Trigger any asynchronous allocation of any CIDR labels we may be selecting
+	// This will cause any endpoints waiting on this to be regenerated twice in
+	// quick succession, but will reduce load on the ipcache.
+	prefixes := newIDSel.selectedCIDRs()
+	if len(prefixes) > 0 {
+		log.WithField("prefixes", prefixes).Debug("inserting CIDR selector prefixes ")
+		sc.ipcache.UpsertPrefixes(prefixes, source.Generated, sc.ipcacheResource(newIDSel.String()))
```

Comment:
This call will eventually attempt to lock the SelectorCache from inside the ipcache, while we are currently holding the SelectorCache lock at this point. 

```
TriggerLabelInjection -> InjectLabels -> UpdatePolicyMaps -> UpdateIdentities
```

Worst case, I believe doing this can cause deadlocks between SelectorCache and ipcache. Otherwise, it may cause double endpoint regeneration. If it doesn't case double endpoint regeneration, then we'd likely be in the case where Cilium has completed "policy recalculation" for any given endpoint, but all the state has not been synced, aka race condition.

Reply:
> we'd likely be in the case where Cilium has completed "policy recalculation" for any given endpoint, but all the state has not been synced, aka race condition.

To elaborate on the above, when a policy containing CIDR rules are applied and a new endpoint arrives on the node, the endpoint goes through initial regeneration. It will trigger the above code path to `UpsertPrefixes` and continues on. This trigger queues the ipcache operation to be done **later**. When the endpoint regeneration completes, it will **not** have actually implemented the CIDR policy, as that comes when the second regeneration completes (what was queued just before, aka "double regeneration"). 

This is problematic because endpoint regeneration includes saying "endpoint now implements policy revision X", but this double regeneration is breaks that. Anything downstream / depending on policy revision will actually see temporary drops briefly until the second regeneration completes. In other words, the endpoint only implements policy X when the second regeneration completes. 

One example of something downstream of the policy revision is how CI system for policy tests, both in the Cilium CLI and in the Ginkgo suite.


To fix this, I think we need synchronization mechanism for ipcache upsert, which is something that has been discussed in context of the [ipcache refactor](https://github.com/cilium/cilium/issues/21142), specifically for the FQDN piece.

Reply:
Let's say we implement the above sync mechanism, the next concern I have is what assumptions might we be breaking by having the selectorcache call into the ipcache, i.e. deadlocks, etc.